File: dynabuffers-java/src/main/kotlin/dynabuffers/ast/datatype/MapType.kt
package dynabuffers.ast.datatype

import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import dynabuffers.ast.datatype.ArrayType.ArrayTypeOptions
import dynabuffers.ast.datatype.StringType.StringTypeOptions
import java.nio.ByteBuffer
import java.nio.charset.Charset

class MapType(val options: MapTypeOptions) : IType, ISerializable {

    private val stringType = StringType(StringTypeOptions(options.charset))
    private val booleanType = BooleanType()
    private val byteType = ByteType()
    private val floatType = FloatType()
    private val intType = IntType()
    private val longType = LongType()
    private val shortType = ShortType()
    private val nullType = NullType()

    override fun size(value: Any?, registry: IRegistry): Int {
        var size = 2
        (value as Map<*, *>)
                .filter { e -> e.key != null }
                .forEach { (k, v) ->
                    val keyType = getKeyType(k!!)

                    size += 4
                    size += keyType.size(k, registry)

                    if (v != null) {
                        val valType = getValType(v)
                        size += valType.size(v, registry)
                    }
                }
        return size
    }

    override fun serialize(value: Any?, buffer: ByteBuffer, registry: IRegistry) {
        buffer.putShort((value as Map<*, *>).size.toShort())
        value
                .filter { e -> e.key != null }
                .forEach { (k, v) ->
                    val keyType = getKeyType(k!!)
                    val valType = getValType(v)

                    val keySize = keyType.size(k, registry)
                    val valSize = valType.size(v, registry)

                    val ordinal = typeToOrdinal(v)
                    buffer.putInt(mergeHeader(ordinal, keySize + valSize))
                    keyType.serialize(k, buffer, registry)
                    valType.serialize(v, buffer, registry)
                }
    }

    override fun deserialize(buffer: ByteBuffer, registry: IRegistry): Any {
        val map = HashMap<String, Any?>()
        for (i in 0 until buffer.short) {
            val header = buffer.int
            val (type, length) = unmergeHeader(header)

            val keyType = stringType
            val valType = ordinalToType(type)

            val rawValue = ByteArray(length)
            buffer.get(rawValue)

            val entry = ByteBuffer.wrap(rawValue)

            // entry key deserialization
            val k = keyType.deserialize(entry, registry)
            // entry val deserialization
            val v = valType.deserialize(entry, registry)

            map[k as String] = v
        }
        return map
    }

    private fun getValType(obj: Any?): ISerializable = when (obj) {
        is String -> stringType
        is Map<*, *> -> this
        is Boolean -> booleanType
        is Byte -> byteType
        is Float -> floatType
        is Int -> intType
        is Long -> longType
        is Short -> shortType
        is Array<*> -> ArrayType(ArrayTypeOptions(getValType(obj.firstOrNull() ?: "")))
        is Collection<*> -> ArrayType(ArrayTypeOptions(getValType(obj.firstOrNull() ?: "")))
        null -> nullType
        else -> throw IllegalArgumentException("cannot handle value $obj")
    }

    private fun getKeyType(obj: Any): ISerializable = when (obj) {
        is String -> stringType
        else -> throw IllegalArgumentException("cannot handle $obj")
    }

    private fun typeToOrdinal(obj: Any?): Byte = when (obj) {
        is String -> 0
        is Boolean -> 10
        is Byte -> 20
        is Float -> 30
        is Int -> 40
        is Long -> 50
        is Short -> 60
        is Map<*, *> -> 70
        is Array<*> -> (80 + typeToOrdinal(obj.firstOrNull() ?: "") / 10).toByte()
        is Collection<*> -> (80 + typeToOrdinal(obj.firstOrNull() ?: "") / 10).toByte()
        null -> 90
        else -> throw IllegalArgumentException("cannot handle value $obj")
    }

    private fun ordinalToType(ordinal: Byte): ISerializable = when (ordinal.toInt()) {
        0 -> stringType
        10 -> booleanType
        20 -> byteType
        30 -> floatType
        40 -> intType
        50 -> longType
        60 -> shortType
        70 -> this
        80 -> ArrayType(ArrayTypeOptions(StringType(StringTypeOptions(options.charset))))
        81 -> ArrayType(ArrayTypeOptions(BooleanType()))
        82 -> ArrayType(ArrayTypeOptions(ByteType()))
        83 -> ArrayType(ArrayTypeOptions(FloatType()))
        84 -> ArrayType(ArrayTypeOptions(IntType()))
        85 -> ArrayType(ArrayTypeOptions(LongType()))
        86 -> ArrayType(ArrayTypeOptions(ShortType()))
        87 -> ArrayType(ArrayTypeOptions(this))
        90 -> nullType
        else -> throw IllegalArgumentException("cannot handle ordinal $ordinal")
    }

    private fun mergeHeader(a: Byte, b: Int): Int {
        val right = 0xFFFFFF
        return a.toInt().shl(24).or(b.and(right))
    }

    private fun unmergeHeader(c: Int): Pair<Byte, Int> {
        val right = 0xFFFFFF

        val aBack = c.shr(24).toByte();
        val bBack = c.and(right)

        return Pair(aBack, bBack)
    }

    data class MapTypeOptions(val charset: Charset)

}

File: dynabuffers-python/dynabuffers/ast/datatype/MapType.py
from typing import Dict, Any

from dynabuffers.api.ISerializable import ISerializable, ByteBuffer
from dynabuffers.ast.datatype.ArrayType import ArrayType, ArrayTypeOptions
from dynabuffers.ast.datatype.BooleanType import BooleanType
from dynabuffers.ast.datatype.ByteType import ByteType
from dynabuffers.ast.datatype.FloatType import FloatType
from dynabuffers.ast.datatype.IntType import IntType
from dynabuffers.ast.datatype.LongType import LongType
from dynabuffers.ast.datatype.ShortType import ShortType
from dynabuffers.ast.datatype.StringType import StringType, StringTypeOptions


class MapTypeOptions:
    def __init__(self, charset: str):
        self.charset = charset


class MapType(ISerializable):

    def __init__(self, options: MapTypeOptions):
        self.options = options

    def size(self, value, registry):
        size = 2
        for key in value:
            if key is None or value[key] is None:
                continue

            key_type = self._get_key_type(key)
            val_type = self._get_val_type(value[key])

            size += 4
            size += key_type.size(key, registry)
            size += val_type.size(value[key], registry)

        return size

    def serialize(self, value, buffer: ByteBuffer, registry):
        buffer.putShort(len(value))
        for key in value:
            if key is None or value[key] is None:
                continue

            key_type = self._get_key_type(key)
            val_type = self._get_val_type(value[key])

            key_size = key_type.size(key, registry)
            val_size = val_type.size(value[key], registry)

            ordinal = self._type_to_ordinal(value[key])
            buffer.putInt(self.merge_header(ordinal, key_size + val_size))
            key_type.serialize(key, buffer, registry)
            val_type.serialize(value[key], buffer, registry)

    def deserialize(self, buffer: ByteBuffer, registry):
        result:Dict[str, Any] = dict()

        counter = 0
        amount = buffer.getShort()
        while counter < amount:
            header = buffer.getInt()
            (type, length) = self.unmerge_header(header)

            key_type = StringType(StringTypeOptions(self.options.charset))
            val_type = self._ordinal_to_type(type)

            raw_value = buffer.get(length)
            entry = ByteBuffer.wrap(raw_value)

            k = key_type.deserialize(entry, registry)
            v = val_type.deserialize(entry, registry)

            result[k] = v
            counter += 1

        return result

    def _get_val_type(self, obj) -> ISerializable:
        if isinstance(obj, str):
            return StringType(StringTypeOptions(self.options.charset))
        if isinstance(obj, dict):
            return self
        if isinstance(obj, bool):
            return BooleanType()
        if isinstance(obj, float):
            return FloatType()
        if isinstance(obj, int):
            return IntType()
        if isinstance(obj, bytes):
            return ByteType()
        if isinstance(obj, list):
            return ArrayType(ArrayTypeOptions(self._get_val_type(obj[0])))
        raise NameError("cannot handle value " + str(obj))

    def _get_key_type(self, obj) -> ISerializable:
        if isinstance(obj, str):
            return StringType(StringTypeOptions(self.options.charset))
        raise NameError("cannot handle value " + str(obj))

    def _type_to_ordinal(self, obj):
        if isinstance(obj, str):
            return 0
        if isinstance(obj, bool):
            return 10
        if isinstance(obj, bytes) and len(obj) == 1:
            return 20
        if isinstance(obj, float):
            return 30
        if isinstance(obj, int):
            return 40
        if isinstance(obj, bytes) and len(obj) == 8:
            return 50
        if isinstance(obj, bytes) and len(obj) == 2:
            return 60
        if isinstance(obj, dict):
            return 70
        if isinstance(obj, list):
            return 80 + self._type_to_ordinal(obj[0])

        raise NameError("cannot handle value " + str(obj))

    def _ordinal_to_type(self, obj):
        if obj == 0:
            return StringType(StringTypeOptions(self.options.charset))
        if obj == 10:
            return BooleanType()
        if obj == 20:
            return ByteType()
        if obj == 30:
            return FloatType()
        if obj == 40:
            return IntType()
        if obj == 50:
            return LongType()
        if obj == 60:
            return ShortType()
        if obj == 70:
            return self
        if obj == 80:
            return ArrayType(ArrayTypeOptions(StringType(StringTypeOptions(self.options.charset))))
        if obj == 81:
            return ArrayType(ArrayTypeOptions(BooleanType()))
        if obj == 82:
            return ArrayType(ArrayTypeOptions(ByteType()))
        if obj == 83:
            return ArrayType(ArrayTypeOptions(FloatType()))
        if obj == 84:
            return ArrayType(ArrayTypeOptions(IntType()))
        if obj == 85:
            return ArrayType(ArrayTypeOptions(LongType()))
        if obj == 86:
            return ArrayType(ArrayTypeOptions(ShortType()))
        if obj == 87:
            return ArrayType(ArrayTypeOptions(self))

        raise NameError("cannot handle value " + str(obj))

    def merge_header(self, a:int, b:int) -> int:
        right = 0xFFFFFF
        return a << 24 | (b & right)

    def unmerge_header(self, c:int) -> (int, int):
        right = 0xFFFFFF
        return c >> 24, c & right
